Add Iterators

iwburns opened the pull request:
Fixes #40

iwburns commented:
Still need to add docs and doc tests for this

iwburns commented:
@Drakulix if you don't mind I'd love to get your thoughts on my approach here.  It's still a WiP, of course, but any thoughts would be appreciated.

Drakulix commented on src/tree/iterators.rs:
```diff
@@ -0,0 +1,252 @@
+
+                self.node_id = Some(parent_id.clone());
+
+                return Some(parent);
```
Spacing looks weird here.

iwburns replied:
I agree, will fix that shortly.

Drakulix commented on src/tree/iterators.rs:
```diff
@@ -0,0 +1,252 @@
+impl<'a, T> IteratorNew<'a, T, ChildrenIds<'a>> for ChildrenIds<'a> {
+    fn new(tree: &'a Tree<T>, node_id: NodeId) -> ChildrenIds<'a> {
+        ChildrenIds { child_ids: tree.get_unsafe(&node_id).children().as_slice().iter() }
+    }
```
Couldn't you just return the `Iterator` directly?

iwburns replied:
No particular reason to be honest.

So (if I'm understanding you correctly), basically the other way to do this would be to have
`tree.children_ids(...)`
just do this:
`return self.get_unsafe(...).children().as_slice().iter()`
(or whatever the equivalent code that type-checks properly would be) and then not have a `ChildrenIds` struct at all, correct?

Drakulix replied:
Yep.
What you really want here is this nightly feature: https://github.com/rust-lang/rfcs/pull/1522

But since there is no real reason, this crate should not work on stable, on second thought, it might be better to hide the underlying Iterator, so you can more easily change it, until RFC 1522 has been stabilized.

iwburns replied:
I think I see what you're saying here, but I'm not sure I'm following how I would
>hide the underlying Iterator, so you can more easily change it, until RFC 1522 has been stabilized.

Do you mind clarifying?

Drakulix replied:
Hiding would be essentially what you do now. Not exposing the type but instead wrapping it in your struct.
I basically changed my mind, thinking about what guarantees we want to make in the API.
You basically want to express, that we return an `Iterator`, not what kind of `Iterator`. Hiding it by wrapping makes it easier to change it later without the need to break the API.

iwburns replied:
Ah, I see what you mean.  Sorry for the confusion, and yes I agree!

Drakulix commented on src/tree/mod.rs:
```diff
@@ -1134,6 +1248,10 @@ impl<T> Tree<T> {
+trait IteratorNew<'a, T, I> {
+    fn new(tree: &'a Tree<T>, node_id: NodeId) -> I;
+}
+
```
Any particular reason, why you have this trait? Any reason not to make these `new` functions "normal" struct methods?

iwburns replied:
If I'm not mistaken, this is the only way to make sure that the `new(...)` function is available the `tree` module, but not have it available any higher up than that.

Essentially I just wanted to be extra safe and not allow the library user to call `new(...)` on anything in the `iterators` module directly and this is the only pattern I could think of that would provide that behavior.

Drakulix replied:
Because the Iterators are in a lower module, I think you would not need to make them public.
But of course, if this is about visibility, it is understandable.

iwburns replied:
I'll see if I can re-visit this so that I don't have to do the trait workaround.  I just don't want them public from the library perspective.  Internally, I don't really care where they are (although I was trying to keep them in their own file because the tree module is getting large).

Drakulix replied:
I would try if making them non-public "normal" methods does compile and otherwise just revert back to this solution.

iwburns commented:
Yeah, I agree it seems odd that you have to ask the `Tree` for these iterators, but I'd have to think through the implications of having `Node`s know which `Tree` they belong to.

Drakulix commented:
> Yeah, I agree it seems odd that you have to ask the Tree for these iterators, but I'd have to think through the implications of having Nodes know which Tree they belong to.

Yes, this would need to be another issue. Just wanted to bring that up, instead of completely ignoring it.

iwburns commented on src/tree/iterators.rs:
```diff
@@ -0,0 +1,401 @@
+            data: Vec::with_capacity(tree.nodes.capacity()),
+        };
+
+        traversal.process_nodes(node_id);
```
@Drakulix I'm not 100% happy with this because it basically forces all of the computation to be run at the beginning, regardless of whether or not the user ends up using the whole iterator.

However, I couldn't come up with a good way to do this incrementally as next is called. (I found a way for PreOrder but PostOrder has stumped me). Do you have any thoughts on this?

Drakulix replied:
It's certainly tricky, but I think not impossible, but maybe not worth the effort.

I guess you would need to store a stack inside the Iterator. Entries would be a node + the current index of its children and then start with index 0 and the root node. Then build up the stack on a `next` call until you have the next element and so on. you are done with a node and pop it from the stack, when the index would go out of bounds.

That *should* be the minimal amount of state required. you could probably also just store the current id + index, but then you would need to back-track via `parent()` and find out the child's position and that sounds unnecessary to me.

iwburns replied:
Yeah, good points.  That would probably be the way to go.  I'm probably going to leave it for now and create an issue for later. Thanks for taking a look!

iwburns commented:
@Drakulix I know it's been a while, but I'd like your opinion again if you have the time.

I'm mostly happy with this as a v1.1.0 release, but I'm wondering if I should add a Level-Order Traversal as well.  It seems like it might be useful enough to include, right?

Also, all of the traversal iterators currently iterate directly over the `Node`s themselves, but I'm wondering if it would be nice to also have a `NodeId` version of each of them (just like the child and ancestor iterators).

Drakulix commented:
Looks good to me, except for that `process_node` call we already talked about. But I would most likely go the easy route as well, this is difficult to optimize and get right.

> I'm mostly happy with this as a v1.1.0 release, but I'm wondering if I should add a Level-Order Traversal as well. It seems like it might be useful enough to include, right?

I even think, I have a use case for this in my window manager. So yes, it is definitely useful, but if it takes too much time now, I would not hold back 1.1.0 for too long and just push that to 1.2.0.

> Also, all of the traversal iterators currently iterate directly over the Nodes themselves, but I'm wondering if it would be nice to also have a NodeId version of each of them (just like the child and ancestor iterators).

I cannot think of many use cases for this and generally this should be avoided, because you will be dealing with `Id`s and therefor a bunch of `unwrap`s or `except`s again. Having only `Node` iterators is a feature not a short-coming in my eyes.

iwburns commented:
>Looks good to me, except for that process_node call we already talked about. But I would most likely go the easy route as well, this is difficult to optimize and get right.

Yeah, I'd love to get it right _and_ fast the first time, but I'd rather not deal with it right now (I've let this sit here much longer than I wanted to originally).  Plus it should be something that can be optimized later without breaking anything.

>I even think, I have a use case for this in my window manager. So yes, it is definitely useful, but if it takes too much time now, I would not hold back 1.1.0 for too long and just push that to 1.2.0.

It shouldn't be too hard to do, but the implementation will be a naive one (similar to the `process_node` situation) for now.  I'm fine with leaving those as future optimizations so that I can get this functionality out there sooner rather than later.

>I cannot think of many use cases for this and generally this should be avoided, because you will be dealing with Ids and therefor a bunch of unwraps or excepts again. Having only Node iterators is a feature not a short-coming in my eyes.

I mostly agree, but I'm worried there are use-cases out there that might require `NodeId` iterators over `Node` iterators that I'm just not aware of.  Maybe that's not something that I should be worrying too much over though.

Drakulix commented:
> I mostly agree, but I'm worried there are use-cases out there that might require NodeId iterators instead of Node iterators that I'm just not aware of. Maybe that's not something that I should be worrying too much over though.

Exactly. In the worst case you get a feature request half a year later.

iwburns commented:
@Drakulix I've switched things up a bit (nothing too crazy, but I forgot I had some changes on a different machine that cleaned up one of the iterators a bit).  I've added the LevelOrderTraversal as well.

Do you mind taking a look when you have a chance?

Drakulix commented:
Just took a look over everything. Looks good to me :)

iwburns commented:
>Including this. This makes switching to this Iterator even more valuable for me.

That's great to hear.  I also simplified the `PreOrderTraversal` a bit just now.  I'm sorry to keep asking, but if you don't mind I'd love another set of eyes on it.  This is the last set of changes that I wanted to make before merging this.

Drakulix commented:
> I also simplified the PreOrderTraversal a bit just now. I'm sorry to keep asking, but if you don't mind I'd love another set of eyes on it.

No worries. Again looks fine. :)
